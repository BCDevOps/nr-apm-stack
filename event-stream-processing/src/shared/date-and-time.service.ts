import {injectable} from 'inversify';
import moment from 'moment';

@injectable()
export class DateAndTimeService {
  now(): moment.Moment {
    return moment();
  }
}
