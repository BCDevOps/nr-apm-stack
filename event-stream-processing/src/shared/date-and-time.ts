export interface DateAndTime {
    now(): moment.Moment;
}
