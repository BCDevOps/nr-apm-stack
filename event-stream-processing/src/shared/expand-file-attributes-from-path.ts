import lodash from 'lodash';
import * as path from 'path';

export function expandFileAttributesFromPath(filePath:string, targetFileObject: any): any {
  const fileName = path.basename(filePath);
  lodash.set(targetFileObject, 'directory', path.dirname(filePath));
  lodash.set(targetFileObject, 'name', fileName);
  const indexOfLastDotInFileName = fileName.lastIndexOf('.');
  if (indexOfLastDotInFileName >0) {
    const fileExt = fileName.substring(indexOfLastDotInFileName+1).trim();
    if (fileExt.length>0) {
      lodash.set(targetFileObject, 'extension', fileName.substring(indexOfLastDotInFileName+1));
    }
  }
}
