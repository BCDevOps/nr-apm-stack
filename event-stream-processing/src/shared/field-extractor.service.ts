import {injectable} from 'inversify';
import * as path from 'path';
import {OsDocument} from '../types/os-document';
import lodash from 'lodash';

@injectable()
export class FieldExtractorService {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  fieldStringToArray(fieldString: string, document: OsDocument): any[] {
    return fieldString.split(',')
      .map((docPath) => {
        if (docPath.startsWith('basename(')) {
          const filePath = lodash.get(document.data, docPath.substring(9, docPath.length - 1), '');
          return path.basename(filePath);
        } else if (docPath === 'kinesis.eventID') {
          // Grab from record as the kinesis fields may not be set
          return document.record.eventID;
        }
        return lodash.get(document.data, docPath, '');
      });
  }
}
