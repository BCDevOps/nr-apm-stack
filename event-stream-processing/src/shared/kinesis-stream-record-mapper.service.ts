import {KinesisStreamRecord} from 'aws-lambda';
import {inject, injectable} from 'inversify';
import lodash from 'lodash';
import {FINGERPRINTS, FINGERPRINT_UNKNOWN} from '../constants/fingerprints';
import {TYPES} from '../inversify.types';
import {OsDocument, OsDocumentData, OsDocumentFingerprint} from '../types/os-document';
import {SubsetService} from './subset.service';

@injectable()
export class KinesisStreamRecordMapperService {
  /**
   * Constructor
   */
  constructor(
    @inject(TYPES.SubsetService) private subset: SubsetService,
  ) {}

  /**
   * Converts a KinesisStreamRecord to an OsDocument
   * @param record The record to convert
   * @returns
   */
  public toOpensearchDocument(record: KinesisStreamRecord): OsDocument {
    const data = JSON.parse(Buffer.from(record.kinesis.data, 'base64').toString('utf8'));
    return {
      fingerprint: FINGERPRINT_UNKNOWN,
      id: null,
      index: null,
      type: '_doc',
      data,
      record,
      error: null,
    };
  }

  /**
   * Matches the document to a fingerprint and modifies the data with any defaults from the fingerprint.
   * @param document The document to modify
   * @returns
   */
  public toFingerprintedDocument(document: OsDocument): OsDocument {
    const fingerprint = this.fingerprintData(document.data);
    lodash.defaultsDeep(document.data, fingerprint.dataDefaults);
    document.fingerprint = fingerprint;
    return document;
  }

  public fingerprintData(data: OsDocumentData): OsDocumentFingerprint {
    return FINGERPRINTS.find((fp) => {
      if (fp.fingerprint === null) {
        return fp;
      }
      if (this.subset.isSubset(data, fp.fingerprint)) {
        return fp;
      }
    }) as OsDocumentFingerprint; // Will always match last "unknown" fingerprint
  }
}
