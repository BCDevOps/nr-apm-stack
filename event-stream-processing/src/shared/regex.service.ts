import lodash from 'lodash';
import {LoggerService} from '../util/logger.service';
import {OsDocument} from '../types/os-document';
import {injectable, inject} from 'inversify';
import {TYPES} from '../inversify.types';

const underscoreReplaceRegex = /__/g;

/**
 * Service for parsing a message field into fields using regex
 */
@injectable()
export class RegexService {
  /**
   * Constructor
   * @param logger
   */
  constructor(
    @inject(TYPES.LoggerService) private logger: LoggerService,
  ) {}

  /**
   * Parse field into fields based on regex
   * @param document The document to modify
   * @param field The field to parse using the regex
   * @param regexArr An array of regex to test for a match
   * @param skipDash Do not add fields where the value is a dash
   */
  applyRegex(document: OsDocument, field: string, regexArr: RegExp[], skipDash = true): void {
    const fieldValue = lodash.get(document.data, field);
    this.logger.debug(`Parsing ${fieldValue as string}`);
    for (const regex of regexArr) {
      const m = fieldValue.match(regex);
      if (m !== null) {
        for (const gropName of Object.keys(m.groups)) {
          const value = m.groups[gropName];
          if (skipDash && value === '-') {
            // dash is usually a special value that indicates empty/missing
            continue;
          }
          const fieldName = gropName.replace(underscoreReplaceRegex, '.');
          lodash.set(document.data, fieldName, value);
        }
        break;
      }
    }
  }
}
