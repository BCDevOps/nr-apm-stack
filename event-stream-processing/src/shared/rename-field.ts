import lodash from 'lodash';

export function renameField(record: any, fromName: string, toName: string): void {
  const value = lodash.get(record, fromName);
  if (!lodash.isNil(value)) {
    lodash.set(record, toName, value);
  }
  lodash.unset(record, fromName);
}
