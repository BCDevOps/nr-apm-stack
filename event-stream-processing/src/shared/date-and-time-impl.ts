/* istanbul ignore file */
import {injectable} from 'inversify';
import * as moment from 'moment';
import {DateAndTime} from './date-and-time';

@injectable()
export class DateAndTimeImpl implements DateAndTime {
  now(): moment.Moment {
    return moment();
  }
}
