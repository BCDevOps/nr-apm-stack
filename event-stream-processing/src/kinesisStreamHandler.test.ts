import {KinesisStreamEvent, Context, KinesisStreamRecord} from 'aws-lambda';
import {create as buildContainer, TYPES} from './inversify.config';
import {KinesisStreamHandler} from './kinesisStreamHandler.isvc';
import {OpenSearch, OpenSearchBulkResult} from './opensearch.isvc';
import * as lodash from 'lodash';
import {AwsHttpClient, HttpBufferedResponse, HttpRequestOptions, HttpResponseWrapper} from './aws-http-client.isvc';
import {APACHE_ACCESS_LOG_EVENT_SIGNATURE} from './parser.apache.svc';
import {Randomizer} from './randomizer.isvc';
import {Logger} from './logger.isvc';
import {LoggerVoidImpl} from './logger-void.svc';
// eslint-disable-next-line max-len
import {APACHE_LOG_COMBINED_FORMAT_1, APACHE_LOG_COMBINED_KEEPALIVE, APACHE_LOG_V1_FORMAT_BAD_1} from './fixture-apache-log';
import {MaxmindAsnLookup, MaxmindCityLookup} from './maxmindLookup.isvc';
import {AsnResponse, CityResponse} from 'maxmind';
import { DateAndTime } from './shared/date-and-time';
import moment = require('moment');

const myContainer = buildContainer();

// eslint-disable-next-line max-len
const message1 = 'v1.0 20120211 "https://testapps.nrs.gov.bc.ca:443" "2001:569:be94:4700:61b4:917e:808:e3c6" [20/Apr/2021:15:10:40 -0700] "POST /int/fncs/activities/details.xhtml HTTP/1.1" 200 2600 bytes 1112 bytes "https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml?activityGuid=0123C676CBAB461E8C6A0A14567A16AC" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36" 133 ms, "TLSv1.2" "ECDHE-RSA-AES256-GCM-SHA384"';
// eslint-disable-next-line max-len,camelcase
const message1_1 = 'v1.0 20120211 "https://testapps.nrs.gov.bc.ca:443" "2001:569:be94:4700:61b4:917e:808:e3c6" [36/Amz/2021:15:10:40 -0700] "POST /int/fncs/activities/details.xhtml HTTP/1.1" 200 2600 bytes 1112 bytes "https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml?activityGuid=0123C676CBAB461E8C6A0A14567A16AC" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36" 133 ms, "TLSv1.2" "ECDHE-RSA-AES256-GCM-SHA384"';
const apacheAccesLogEvent = APACHE_ACCESS_LOG_EVENT_SIGNATURE;
const event1 = Object.assign({message: message1}, apacheAccesLogEvent);
const record1 = {
  kinesis: {
    sequenceNumber: '123',
    data: Buffer.from(JSON.stringify(event1), 'utf8').toString('base64'),
  },
} as any as KinesisStreamRecord;
const record2 = {
  kinesis: {
    sequenceNumber: '456',
    data: Buffer.from(JSON.stringify(event1), 'utf8').toString('base64'),
  },
} as any as KinesisStreamRecord;
const record3 = {
  kinesis: {
    sequenceNumber: '456',
    // eslint-disable-next-line max-len
    data: Buffer.from(JSON.stringify(Object.assign({message: message1_1}, apacheAccesLogEvent)), 'utf8').toString('base64'),
  },
} as any as KinesisStreamRecord;

function mockContext(): Context {
  return {} as any as Context;
}

const geoIpCityLookupVictoria:CityResponse = {
  continent: {code: 'NA', geoname_id: 0, names: {en: 'North America'}},
  country: {iso_code: 'CA', geoname_id: 0, names: {en: 'Canada'}},
  subdivisions: [{geoname_id: 0, iso_code: 'BC', names: {en: 'British Columbia'}}],
  city: {geoname_id: 1, names: {en: 'Victoria'}},
  location: {latitude: 0, longitude: 0, accuracy_radius: 0, time_zone: 'America/Vancouver'},
  postal: {code: 'ABC-123'},
};

const geoIpAsnLookupBcGov:AsnResponse = {
  autonomous_system_number: 123456,
  autonomous_system_organization: 'TEST',
};

beforeEach(() => {
  myContainer.snapshot();
  myContainer.rebind<Randomizer>(TYPES.Randomizer).toConstantValue({randomBytes: ()=>{
    return Buffer.from([0x62, 0x75, 0x66, 0x66, 0x65, 0x72]);
  }});
  myContainer.rebind<Logger>(TYPES.Logger).to(LoggerVoidImpl);
  myContainer.rebind<MaxmindCityLookup>(TYPES.MaxmindCityLookup).toConstantValue({
    lookup: ():CityResponse => {
      return geoIpCityLookupVictoria;
    },
  });
  myContainer.rebind<MaxmindAsnLookup>(TYPES.MaxmindAsnLookup).toConstantValue({
    lookup: ():AsnResponse => {
      return geoIpAsnLookupBcGov;
    },
  });
  myContainer.rebind<DateAndTime>(TYPES.DateAndTime).toConstantValue({
    now: (): moment.Moment => {
      return moment('2021-05-26T18:47:40.314-07:00');
    },
  });
});

afterEach(() => {
  myContainer.restore();
});

test('handler - success', async () => {
  const ctx: Context = mockContext();
  const event: KinesisStreamEvent = {Records: [record1]};
  // myContainer.unbind(TYPES.OpenSearch)
  const openSearchMock = {
    bulk: (): Promise<OpenSearchBulkResult> =>{
      return Promise.resolve({success: true, errors: []});
    },
  };
  // myContainer.rebind()
  myContainer.rebind<OpenSearch>(TYPES.OpenSearch).toConstantValue(openSearchMock);
  expect(event).toMatchSnapshot('895ca8a6-0788-41ba-b921-9edbbd3212dc');
  // eslint-disable-next-line max-len
  await expect(myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler).handle(event, ctx)).resolves.not.toThrow();
});

test('handler - all errors', async () => {
  const ctx: Context = mockContext();
  const event: KinesisStreamEvent = {Records: [record1, record2]};
  // myContainer.unbind(TYPES.OpenSearch)
  const openSearchMock = {
    bulk: (documents: any[]): Promise<OpenSearchBulkResult> =>{
      // return Promise.resolve({errors:false, items:[]})
      return Promise.resolve({success: false, errors: documents});
    },
  };
  // myContainer.rebind()
  myContainer.rebind<OpenSearch>(TYPES.OpenSearch).toConstantValue(openSearchMock);
  expect(event).toMatchSnapshot('d52cf91b-50a5-4f26-9d3e-c312778360e0');
  const result = await myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler).handle(event, ctx);
  expect(result).toBeUndefined();
});

test('handler - partial error', async () => {
  const ctx: Context = mockContext();
  // note: records need to be cloned so they don't interfer with each other and other tests
  const event: KinesisStreamEvent = {Records: [
    lodash.set(JSON.parse(JSON.stringify(record1)), 'kinesis.sequenceNumber', '1'),
    lodash.set(JSON.parse(JSON.stringify(record3)), 'kinesis.sequenceNumber', '2'),
    lodash.set(JSON.parse(JSON.stringify(record2)), 'kinesis.sequenceNumber', '3'),
    lodash.set(JSON.parse(JSON.stringify(record1)), 'kinesis.sequenceNumber', '4'),
    lodash.set(JSON.parse(JSON.stringify(record1)), 'kinesis.sequenceNumber', '5'),
  ]};
  const awsHttpClientMock = {
    executeSignedHttpRequest: (httpRequestParams: HttpRequestOptions): Promise<HttpResponseWrapper> => {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const items = httpRequestParams.body.trim().split('\n');
      const response = {body: {errors: false, items: [] as any[]}};
      for (let index = 0; index < items.length; index+=2) {
        const item = JSON.parse(items[index]);
        const item2:any = {create: {_id: item.create._id}};
        if (response.body.items.length % 2 === 0) {
          response.body.errors = true;
          item2.create.error = {};
        }
        response.body.items.push(item2);
      }
      return Promise.resolve({response} as HttpResponseWrapper);
    },
    waitAndReturnResponseBody: (value: HttpResponseWrapper): Promise<HttpBufferedResponse> => {
      return Promise.resolve({statusCode: 200, body: JSON.stringify(value.response.body)} as HttpBufferedResponse);
    },
  } as AwsHttpClient;
  // myContainer.rebind()
  myContainer.rebind<AwsHttpClient>(TYPES.AwsHttpClient).toConstantValue(awsHttpClientMock);
  // await expect(event).toMatchSnapshot('d52cf91b-50a5-4f26-9d3e-c312778360e0')
  const result = await myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler).handle(event, ctx);
  expect(result).toBeUndefined();
});

for (let index = 1; index <= 2; index++) {
  test(`transformToElasticCommonSchema1.${index}`, async () => {
    const event:KinesisStreamEvent = {
      Records: [
        record1,
      ],
    };
    // eslint-disable-next-line max-len
    const events = await myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler).transformToElasticCommonSchema(event);
    expect(events).toHaveLength(1);
    expect(events[0]).toHaveProperty('_index');
    expect(events[0]).toHaveProperty('client.ip');
    expect(events[0]).toHaveProperty('client.geo');
  });
}

test('e2e - Apache combined format', async () => {
  const logEvent = Object.assign({message: APACHE_LOG_COMBINED_FORMAT_1}, APACHE_ACCESS_LOG_EVENT_SIGNATURE);
  const record = {
    kinesis: {
      sequenceNumber: '123',
      data: Buffer.from(JSON.stringify(logEvent), 'utf8').toString('base64'),
    },
  } as any as KinesisStreamRecord;
  const event:KinesisStreamEvent = {Records: [record]};
  // eslint-disable-next-line max-len
  const documents = await myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler).transformToElasticCommonSchema(event);
  expect(documents).toHaveLength(1);
  for (const document of documents) {
    expect(document).toHaveProperty('url.uri');
    expect(document).not.toHaveProperty('url.scheme');
    expect(document).not.toHaveProperty('url.domain');
    expect(document).not.toHaveProperty('url.port');
    expect(document).not.toHaveProperty('url.full');
  }
  expect(documents).toMatchSnapshot('3ac5c5a8-241c-4501-9f25-4981a34e5d02');
});
test('e2e - v1 bad 1', async () => {
  const logEvent = Object.assign({message: APACHE_LOG_V1_FORMAT_BAD_1}, APACHE_ACCESS_LOG_EVENT_SIGNATURE);
  const record = {
    kinesis: {
      sequenceNumber: '123',
      data: Buffer.from(JSON.stringify(logEvent), 'utf8').toString('base64'),
    },
  } as any as KinesisStreamRecord;
  const event:KinesisStreamEvent = {Records: [record]};
  // eslint-disable-next-line max-len
  const documents = await myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler).transformToElasticCommonSchema(event);
  expect(documents).toHaveLength(1);
  for (const document of documents) {
    expect(document).toHaveProperty('http.request.line');
    expect(document).toHaveProperty('http.request.method');
    expect(document).toHaveProperty('http.request.bytes');
    expect(document).toHaveProperty('http.response.bytes');
    expect(document).toHaveProperty('http.response.status_code');
    expect(document).toHaveProperty('http.version');
    expect(document).toHaveProperty('url.uri');
    expect(document).toHaveProperty('url.scheme', 'http');
    expect(document).toHaveProperty('url.domain');
    expect(document).toHaveProperty('url.port');
    expect(document).toHaveProperty('url.full');
  }
  // await expect(documents).toMatchSnapshot('3ac5c5a8-241c-4501-9f25-4981a34e5d02')
});

test('e2e - combined - keepalive', async () => {
  const logEvent = Object.assign({message: APACHE_LOG_COMBINED_KEEPALIVE.message}, APACHE_ACCESS_LOG_EVENT_SIGNATURE);
  const record = {
    kinesis: {
      sequenceNumber: '123',
      data: Buffer.from(JSON.stringify(logEvent), 'utf8').toString('base64'),
    },
  } as any as KinesisStreamRecord;
  const event:KinesisStreamEvent = {Records: [record]};
  // eslint-disable-next-line max-len
  const documents = await myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler).transformToElasticCommonSchema(event);
  expect(documents).toHaveLength(1);
  for (const document of documents) {
    expect(document).toHaveProperty('http.request.line', 'GET /keepalive.html HTTP/1.1');
    expect(document).toHaveProperty('http.request.method');
    expect(document).toHaveProperty('http.response.bytes');
    expect(document).toHaveProperty('http.response.status_code');
    expect(document).toHaveProperty('http.version');
    expect(document).toHaveProperty('url.uri');
    expect(document).not.toHaveProperty('url.scheme');
    expect(document).not.toHaveProperty('url.domain');
    expect(document).not.toHaveProperty('url.port');
    expect(document).not.toHaveProperty('url.full');
  }
  // await expect(documents).toMatchSnapshot('3ac5c5a8-241c-4501-9f25-4981a34e5d02')
});
