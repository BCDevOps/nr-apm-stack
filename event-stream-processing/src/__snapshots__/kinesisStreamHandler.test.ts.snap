// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e - Apache combined format: 3ac5c5a8-241c-4501-9f25-4981a34e5d02 1`] = `
Array [
  Object {
    "@timestamp": "2020-06-05T07:23:23.000-07:00",
    "_id": "3132332e363237353636363636353732",
    "_index": "nrm-logs-access-2020.06.05",
    "event": Object {
      "category": "web",
      "dataset": "apache.access",
      "duration": "24",
      "ingested": "2021-05-26T18:47:40.314-07:00",
      "kind": "event",
      "outcome": "success",
    },
    "http": Object {
      "request": Object {
        "line": "GET /ext/raad3/map?execution=e1s1&_eventId=agreed HTTP/1.1",
        "method": "GET",
        "referrer": Object {
          "domain": "apps.nrs.gov.bc.ca",
          "file": Object {
            "directory": "/ext/raad3",
            "name": "map",
          },
          "full": "https://apps.nrs.gov.bc.ca/ext/raad3/map?execution=e1s1",
          "original": "https://apps.nrs.gov.bc.ca/ext/raad3/map?execution=e1s1",
          "path": "/ext/raad3/map",
          "port": "443",
          "query": "execution=e1s1",
          "query_k": Array [
            "execution",
          ],
          "query_kv": Array [
            "execution=e1s1",
          ],
          "query_v": Array [
            "e1s1",
          ],
          "scheme": "https",
          "uri": "//apps.nrs.gov.bc.ca/ext/raad3/map",
        },
      },
      "response": Object {
        "status_code": "302",
      },
      "version": "1.1",
    },
    "message": "66.183.191.120 - - [05/Jun/2020:07:23:23 -0700] \\"GET /ext/raad3/map?execution=e1s1&_eventId=agreed HTTP/1.1\\" 302 - \\"https://apps.nrs.gov.bc.ca/ext/raad3/map?execution=e1s1\\" \\"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:77.0) Gecko/20100101 Firefox/77.0\\" 24",
    "source": Object {
      "as": Object {
        "number": 852,
        "organization": Object {
          "name": "ASN852",
        },
      },
      "geo": Object {
        "city_name": "Cranbrook",
        "continent_code": "NA",
        "continent_name": "North America",
        "country_iso_code": "CA",
        "country_name": "Canada",
        "location": Object {
          "lat": 49.5014,
          "lon": -115.7734,
        },
        "postal_code": "V1C",
        "region_iso_code": "BC",
        "region_name": "British Columbia",
        "timezone": "America/Edmonton",
      },
      "ip": "66.183.191.120",
    },
    "url": Object {
      "file": Object {
        "directory": "/ext/raad3",
        "name": "map",
      },
      "original": "/ext/raad3/map?execution=e1s1&_eventId=agreed",
      "path": "/ext/raad3/map",
      "query": "execution=e1s1&_eventId=agreed",
      "query_k": Array [
        "execution",
        "_eventId",
      ],
      "query_kv": Array [
        "execution=e1s1",
        "_eventId=agreed",
      ],
      "query_v": Array [
        "e1s1",
        "agreed",
      ],
      "uri": "//localhost/ext/raad3/map",
    },
    "user_agent": Object {
      "name": "Firefox",
      "original": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:77.0) Gecko/20100101 Firefox/77.0",
      "os": Object {
        "name": "Windows",
        "version": "10",
      },
      "version": "77.0",
    },
  },
]
`;

exports[`handler - all errors: d52cf91b-50a5-4f26-9d3e-c312778360e0 1`] = `
Object {
  "Records": Array [
    Object {
      "kinesis": Object {
        "data": "eyJtZXNzYWdlIjoidjEuMCAyMDEyMDIxMSBcImh0dHBzOi8vdGVzdGFwcHMubnJzLmdvdi5iYy5jYTo0NDNcIiBcIjIwMDE6NTY5OmJlOTQ6NDcwMDo2MWI0OjkxN2U6ODA4OmUzYzZcIiBbMjAvQXByLzIwMjE6MTU6MTA6NDAgLTA3MDBdIFwiUE9TVCAvaW50L2ZuY3MvYWN0aXZpdGllcy9kZXRhaWxzLnhodG1sIEhUVFAvMS4xXCIgMjAwIDI2MDAgYnl0ZXMgMTExMiBieXRlcyBcImh0dHBzOi8vdGVzdGFwcHMubnJzLmdvdi5iYy5jYS9pbnQvZm5jcy9hY3Rpdml0aWVzL2RldGFpbHMueGh0bWw/YWN0aXZpdHlHdWlkPTAxMjNDNjc2Q0JBQjQ2MUU4QzZBMEExNDU2N0ExNkFDXCIgXCJNb3ppbGxhLzUuMCAoV2luZG93cyBOVCAxMC4wOyBXT1c2NCkgQXBwbGVXZWJLaXQvNTM3LjM2IChLSFRNTCwgbGlrZSBHZWNrbykgQ2hyb21lLzg5LjAuNDM4OS4xMjggU2FmYXJpLzUzNy4zNlwiIDEzMyBtcywgXCJUTFN2MS4yXCIgXCJFQ0RIRS1SU0EtQUVTMjU2LUdDTS1TSEEzODRcIiIsImV2ZW50Ijp7ImtpbmQiOiJldmVudCIsImNhdGVnb3J5Ijoid2ViIiwiZGF0YXNldCI6ImFwYWNoZS5hY2Nlc3MiLCJpbmdlc3RlZCI6IjIwMjEtMDUtMjZUMTg6NDc6NDAuMzE0LTA3OjAwIn19",
        "sequenceNumber": "123",
      },
    },
    Object {
      "kinesis": Object {
        "data": "eyJtZXNzYWdlIjoidjEuMCAyMDEyMDIxMSBcImh0dHBzOi8vdGVzdGFwcHMubnJzLmdvdi5iYy5jYTo0NDNcIiBcIjIwMDE6NTY5OmJlOTQ6NDcwMDo2MWI0OjkxN2U6ODA4OmUzYzZcIiBbMjAvQXByLzIwMjE6MTU6MTA6NDAgLTA3MDBdIFwiUE9TVCAvaW50L2ZuY3MvYWN0aXZpdGllcy9kZXRhaWxzLnhodG1sIEhUVFAvMS4xXCIgMjAwIDI2MDAgYnl0ZXMgMTExMiBieXRlcyBcImh0dHBzOi8vdGVzdGFwcHMubnJzLmdvdi5iYy5jYS9pbnQvZm5jcy9hY3Rpdml0aWVzL2RldGFpbHMueGh0bWw/YWN0aXZpdHlHdWlkPTAxMjNDNjc2Q0JBQjQ2MUU4QzZBMEExNDU2N0ExNkFDXCIgXCJNb3ppbGxhLzUuMCAoV2luZG93cyBOVCAxMC4wOyBXT1c2NCkgQXBwbGVXZWJLaXQvNTM3LjM2IChLSFRNTCwgbGlrZSBHZWNrbykgQ2hyb21lLzg5LjAuNDM4OS4xMjggU2FmYXJpLzUzNy4zNlwiIDEzMyBtcywgXCJUTFN2MS4yXCIgXCJFQ0RIRS1SU0EtQUVTMjU2LUdDTS1TSEEzODRcIiIsImV2ZW50Ijp7ImtpbmQiOiJldmVudCIsImNhdGVnb3J5Ijoid2ViIiwiZGF0YXNldCI6ImFwYWNoZS5hY2Nlc3MiLCJpbmdlc3RlZCI6IjIwMjEtMDUtMjZUMTg6NDc6NDAuMzE0LTA3OjAwIn19",
        "sequenceNumber": "456",
      },
    },
  ],
}
`;

exports[`handler - success: 895ca8a6-0788-41ba-b921-9edbbd3212dc 1`] = `
Object {
  "Records": Array [
    Object {
      "kinesis": Object {
        "data": "eyJtZXNzYWdlIjoidjEuMCAyMDEyMDIxMSBcImh0dHBzOi8vdGVzdGFwcHMubnJzLmdvdi5iYy5jYTo0NDNcIiBcIjIwMDE6NTY5OmJlOTQ6NDcwMDo2MWI0OjkxN2U6ODA4OmUzYzZcIiBbMjAvQXByLzIwMjE6MTU6MTA6NDAgLTA3MDBdIFwiUE9TVCAvaW50L2ZuY3MvYWN0aXZpdGllcy9kZXRhaWxzLnhodG1sIEhUVFAvMS4xXCIgMjAwIDI2MDAgYnl0ZXMgMTExMiBieXRlcyBcImh0dHBzOi8vdGVzdGFwcHMubnJzLmdvdi5iYy5jYS9pbnQvZm5jcy9hY3Rpdml0aWVzL2RldGFpbHMueGh0bWw/YWN0aXZpdHlHdWlkPTAxMjNDNjc2Q0JBQjQ2MUU4QzZBMEExNDU2N0ExNkFDXCIgXCJNb3ppbGxhLzUuMCAoV2luZG93cyBOVCAxMC4wOyBXT1c2NCkgQXBwbGVXZWJLaXQvNTM3LjM2IChLSFRNTCwgbGlrZSBHZWNrbykgQ2hyb21lLzg5LjAuNDM4OS4xMjggU2FmYXJpLzUzNy4zNlwiIDEzMyBtcywgXCJUTFN2MS4yXCIgXCJFQ0RIRS1SU0EtQUVTMjU2LUdDTS1TSEEzODRcIiIsImV2ZW50Ijp7ImtpbmQiOiJldmVudCIsImNhdGVnb3J5Ijoid2ViIiwiZGF0YXNldCI6ImFwYWNoZS5hY2Nlc3MiLCJpbmdlc3RlZCI6IjIwMjEtMDUtMjZUMTg6NDc6NDAuMzE0LTA3OjAwIn19",
        "sequenceNumber": "123",
      },
    },
  ],
}
`;
